### DISCUSSION TITLE
Hide parameters with CallerXxxAttribute from public API

### DISCUSSION THREAD

**Discussion by ENikS** [dotnet/roslyn#54940] [Ideas] [answered]
I am implementing a SourceGenerator for [dependency injection](https://github.com/unitycontainer) container. 

In the generator I am visiting the code tree and recording occurrences of `Register` and `Resolve` method calls. Based on the analysis of these calls I am generating appropriate implementations. 

To match these implementations at runtime I need some form of ID. If I add something like:
```c#
public object Resolve(some parameter,
        [CallerFilePath] string sourceFilePath = "",
        [CallerLineNumber] int sourceLineNumber = 0)
```  
it will provide all that I need, but I would have to add it to the public API, defined by interface:

```c#
public interface SomeInterface
{
    public object Resolve(some parameter)
}
```  

I wonder if these compiler generated parameters could be added to the implementation without adding them to the public API? Perhaps a property `GenerateOnly` could be added to the attributes, to indicate that these parameters are generated by the compiler, and are not available to developers?

```c#
public interface SomeInterface
{
    public object Resolve(some parameter,
        [CallerFilePath(GenerateOnly = true)] string sourceFilePath = "",
        [CallerLineNumber(GenerateOnly = true)] int sourceLineNumber = 0)
}
```  

I think this is relatively common problem for Source Generators and should be addressed one way or the other. Perhaps I am missing something and there is a better way to solve this issue?

**Comment by CyrusNajmabadi** [comment, marked as answer]
> it will provide all that I need, but I would have to add it to the public API, defined by interface:

No, these would have to be added to the public API.  The compiler *must* see these in order to know at the callsite that these values should be substituted in.

**Reply by ENikS** [reply]
Correct, it will be added to public API like this:

```c#
public interface SomeInterface
{
    public object Resolve(some parameter,
        [CallerFilePath(CompilerOnly = true)] string sourceFilePath = "",
        [CallerLineNumber(CompilerOnly = true)] int sourceLineNumber = 0)
}
```

Notice the attribute: `[CallerFilePath(CompilerOnly = true)]`
It should indicate to the IDE and the compiler that the parameter is generated and should not be used/supplied by the developer. It should not even be offered by IntelliSense as an option.

**Comment by CyrusNajmabadi** [comment]
Could you clarify *why* you want a change here? 

```c#
ResolveX(some parameter,
        [CallerFilePath] string sourceFilePath = "",
        [CallerLineNumber] int sourceLineNumber = 0)
{
    ResolveY(parameter, sourceFilePath, soruceLineNumber);
}

ResolveY(some parameter,
        [CallerFilePath] string sourceFilePath = "",
        [CallerLineNumber] int sourceLineNumber = 0)
{
}
```

Here the core logic is implemented by ResolveY, and users may call that directly.  However, tehy may also call some of the others, and you'd want this code to forward these values along so that the CallerXXX info is the external caller's info, not something unhelpful like telling you that hte caller was right above you.

**Reply by ENikS** [reply]
In each of these calls `parameter` is different and requires generation of different implementation
```c#
ResolveX(some parameter1,
        [CallerFilePath] string sourceFilePath = "",
        [CallerLineNumber] int sourceLineNumber = 0)
{
    ...
    ResolveY(parameter2, sourceFilePath, soruceLineNumber);
}

ResolveY(some parameter3,
        [CallerFilePath] string sourceFilePath = "",
        [CallerLineNumber] int sourceLineNumber = 0)
{
}

```

This is an original interface (the public API):

```c#
public interface SomeInterface
{
    public object Resolve(some parameter)
}
```

in order for the generator to associate generated code with particular call cite I am modifying the signature (adding location information):

```c#
public interface SomeInterface
{
    public object Resolve(some parameter,
        [CallerFilePath(GenerateOnly = true)] string sourceFilePath = "",
        [CallerLineNumber(GenerateOnly = true)] int sourceLineNumber = 0)
}
```

But I would like the developer to see the same API as before. I want these new parameters to be generated by the compiler and to not be available for override by developer's code. 

Inside `Resolve` implementation is matched with location and executed...

**Comment by mrlacey** [comment]
I found this while looking for similar capabilities.

While looking to add logic that requires the correct Caller information, it could break the application if a consumer provided these values incorrectly.

Ideally, we don't want the developer to even know about the additional parameters. If the compiler could want (or error) if they're explicitly set in code, even better.

I understand that this shouldn't be the default behavior (as the `ResolveX` calling `ResolveY` example demonstrates), but there are scenarios where it would be very useful.

There are some workarounds we'll have to rely upon without this, such as heavily worded documentation and custom analyzers to warn when anyone does try to explicitly set these values.
